site/_posts/2021-09-16-how-to-authenticate-your-react-app.md

```diff
@@ -0,0 +1,1087 @@
+
+## Logging out
+Just like `/login`, we’ll create a `/logout` route to make logging out easily accessible anywhere in our React client:
+
+```bash
+touch server/routes/logout.js
+```
+
+Here is how your filetree should look so far for your `server` directory:
+
+```
+server
+├─routes
+│ ├─login.js
+│ ├─logout.js
+│ ├─oauth-callback.js
+│ └─user.js
+└─index.js
+```
+
+
+In `server/index.js`:
+```js
+// ...
+const logoutRoute = (require('./routes/logout'));
+// ...
+// use routes
+app.use('/user', userRoute);
+app.use('/login', loginRoute);
+app.use('/oauth-callback', oauthCallbackRoute);
+app.use('/logout', logoutRoute);
+
+// ...
+```
+
+Our `/logout` route will wipe the user’s saved login by deleting:
+
+- A browser cookie that connects the browser to the server-side session
+- Our Express server-side session (and the token we store there)
+- FusionAuth’s OAuth session for that user
+
+The first two are accomplished by calling the function `req.session.destroy()`. The last is done by redirecting the user to FusionAuth’s `/oauth2/logout` endpoint.
+
+In `server/routes/logout.js`:
+
+```js
+const express = require('express');
+const router = express.Router();
+const config = require('../../client/src/config');
+
+router.get('/', (req, res) => {
+  // delete the session
+  req.session.destroy();
+
+  // end FusionAuth session
+  res.redirect(`http://localhost:${config.fusionAuthPort}/oauth2/logout?client_id=${config.clientID}`);
+});
+
+module.exports = router;
+```
+
+After cleaning up, FusionAuth will redirect to the React app, because we configured the main entry point of our React app as the "Logout URL" in the FusionAuth application configuration.
+
+Try it out. You can also watch cookies in your browser’s developer tools to see it working.
+
+
+## Troubleshooting
+
+FusionAuth has made troubleshooting easy. If you’ve missed a setting in your OAuth configuration, you will most likely see an error message come back telling you what you’ve missed. Check back through the settings above to make sure you have copied the correct information over from FusionAuth, and that you have saved both the OAuth settings as well as the Attribute Mapping in the miniOrange OAuth Client.
+
+You might find that after a restart or a long time between logins, your FusionAuth instance will log out. When that happens, clicking the login on your login form will add an additional step for you to log into your FusionAuth instance again so that it can access the needed credentials.
+
+## Take it further
+
+Some areas that you may want explore further::
+  - Using PKCE (pronounced 'Pixie') to give an additional layer of security to your React app's login workflows.
+  - Exploring roles and self-registration.
+  - Using an iFrame to have a custom login screen with no redirects.
+
+We're working on a tutorial on iFrame as well as exploring roles in your React app, so keep your eyes peeled for those in the upcoming months!
```

Comment:
Let's remove this. We can always add this back in when we publish a post on iframes.